docs/design.md

```diff
@@ -0,0 +1,785 @@
+# Deckhand Design
+
+## Purpose
+
+Deckhand is a document-based configuration storage service built with
+auditability and validation in mind.
+
+## Essential Functionality
+
+* layering - helps reduce duplication in configuration while maintaining
+  auditability across many sites
+* substitution - provides separation between secret data and other
+  configuration data, while allowing a simple interface for clients
+* revision history - improves auditability and enables services to provide
+  functional validation of a well-defined collection of documents that are
+  meant to operate together
+* validation - allows services to implement and register different kinds of
+  validations and report errors
+
+## Documents
+
+All configuration data is stored entirely as structured documents, for which
+schemas must be registered.
+
+### Document Format
+
+The document format is modeled loosely after Kubernetes practices. The top
+level of each document is a dictionary with 3 keys: `schema`, `metadata`, and
+`data`.
+
+* `schema` - Defines the name of the JSON schema to be used for validation.
+  Must have the form: `<namespace>/<kind>/<version>`, where the meaning of
+  each component is:
+  * `namespace` - Identifies the owner of this type of document. The
+    values `deckhand` and `metadata` are reserved for internal use.
+  * `kind` - Identifies a type of configuration resource in the namespace.
+  * `version` - Describe the version of this resource, e.g. "v1".
+* `metadata` - Defines details that Deckhand will inspect and understand. There
+  are multiple schemas for this section as discussed below. All the various
+  types of metadata include a `name` field which must be unique for each
+  document `schema`.
+* `data` - Data to be validated by the schema described by the `schema`
+  field. Deckhand only interacts with content here as instructed to do so by
+  the `metadata` section. The form of this section is considered to be
+  completely owned by the `namespace` in the `schema`.
+
+#### Document Metadata
+
+There are 3 supported kinds of document metadata. Documents with `Document`
+metadata are the most common, and are used for normal configuration data.
+Documents with `Control` metadata are used to customize the behavior of
+Deckhand. Documents with `Tombstone` metadata are used to delete pre-existing
+documents with either `Document` or `Control` metadata.
+
+##### schema: metadata/Document/v1
+
+This type of metadata allows the following metadata hierarchy:
+
+* `name` - string, required - Unique within a revision for a given `schema`.
+* `storagePolicy` - string, required - Either `cleartext` or `encrypted`. If
+  `encyrpted` is specified, then the `data` section of the document will be
```

Comment:
Regarding: "either a centralized access policy approach or an individual document-level approach could be fine".

Centralized access policy is what all of OpenStack does and is much easier to debug and implement, particularly because ``oslo.policy`` can be leveraged to accomplish that. The advantages are pretty straightforward: 

- easier to integrate/implement than per-document RBAC
- self-documenting
- tried and true implementation of RBAC

Per-document RBAC would require retrieving the access policy from the DB and parsing it to figure out whether a particular read, delete or write operation is allowed. For example, it seems inefficient for a role to have read and write permissions but not delete permissions, so this means we could get 50% of the way through an intricate YAML merge operation only to fail on some delete operation. It might be possible to figure out in advance all the operations needed to completely merge a YAML document, but clearly per-document RBAC is a much more difficult thing to implement and debug. That is, if there is an issue with the access policy itself and not the YAML file, then the error thrown by Deckhand would probably have to return the access policy information in addition to where it failed in the YAML merge procedure just in order for a user to figure out the source of the problem.

On the other hand, per-document RBAC is dynamic and far stricter than a more centralized RBAC approach using ``oslo.policy``.